cpp/core.cc

```diff
@@ -211,48 +214,62 @@ void SignalHandler::Update(int sig_num) {
 }
 
 static List<int>* AllocSignalQueue() {
+  NO_ROOTS_FRAME(FUNC_NAME);  // NewList() does it
   List<int>* ret = NewList<int>();
   ret->reserve(kMaxSignalsInFlight);
   return ret;
 }
 
 List<int>* SignalHandler::TakeSignalQueue() {
+  NO_ROOTS_FRAME(FUNC_NAME);  // AllocSingalQueue() does it
   List<int>* new_queue = AllocSignalQueue();
   List<int>* ret = signal_queue_;
   signal_queue_ = new_queue;
   return ret;
 }
 
 void Sigaction(int sig_num, sighandler_t handler) {
+  NO_ROOTS_FRAME(FUNC_NAME);  // no allocations here
   struct sigaction act = {};
   act.sa_handler = handler;
   assert(sigaction(sig_num, &act, nullptr) == 0);
 }
 
 static void signal_handler(int sig_num) {
-  gSignalHandler.Update(sig_num);
+  assert(gSignalHandler != nullptr);
+  gSignalHandler->Update(sig_num);
 }
 
 void RegisterSignalInterest(int sig_num) {
+  NO_ROOTS_FRAME(FUNC_NAME);  // no allocations here
   struct sigaction act = {};
   act.sa_handler = signal_handler;
   assert(sigaction(sig_num, &act, nullptr) == 0);
 }
 
 List<int>* TakeSignalQueue() {
-  return gSignalHandler.TakeSignalQueue();
+  NO_ROOTS_FRAME(FUNC_NAME);  // SignalHandler::TakeSignalQueue() does it
+  assert(gSignalHandler != nullptr);
+  auto* ret = gSignalHandler->TakeSignalQueue();
```

Comment:
I would just use the original style of `return gSignalHandler...`

Reply:
Ah whoops. This was leftover from some debugging I was doing